core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/MetadataSetupService.java

```diff
@@ -184,12 +164,114 @@ public int saveInstanceMetaData(Stack stack, Iterable<CloudVmMetaDataStatus> clo
         }
     }
 
+    private boolean setupInstanceMetaDataType(boolean primaryIgSelected, Optional<InstanceMetaData> terminatedPrimaryGwWhichShouldBeRestored,
+            InstanceMetaData instanceMetaDataEntry, InstanceGroup ig) {
+        if (instanceMetaDataEntry.getInstanceMetadataType() == null) {
+            if (ig != null) {
+                LOGGER.info("Instance group type: {}", ig.getInstanceGroupType());
+                if (InstanceGroupType.GATEWAY.equals(ig.getInstanceGroupType())) {
+                    primaryIgSelected = setupGatewayInstance(primaryIgSelected, terminatedPrimaryGwWhichShouldBeRestored, instanceMetaDataEntry);
+                } else {
+                    LOGGER.info("Instance is a core instance: {}", instanceMetaDataEntry.getInstanceId());
+                    instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.CORE);
+                }
+            } else {
+                LOGGER.info("Instance group is null, instance will be a core instance: {}", instanceMetaDataEntry.getInstanceId());
+                instanceMetaDataEntry.setInstanceMetadataType(InstanceMetadataType.CORE);
+            }
+        }
+        return primaryIgSelected;
+    }
+
+    private void setupFromCloudInstance(CloudInstance cloudInstance, InstanceMetaData instanceMetaDataEntry) {
+        instanceMetaDataEntry.setSubnetId(cloudInstance.getSubnetId());
+        instanceMetaDataEntry.setAvailabilityZone(cloudInstance.getAvailabilityZone());
+        instanceMetaDataEntry.setInstanceName(cloudInstance.getStringParameter(CloudInstance.INSTANCE_NAME));
+    }
+
+    private void setupFromCloudInstanceMetadata(CloudInstanceMetaData md, InstanceMetaData instanceMetaDataEntry) {
+        instanceMetaDataEntry.setPrivateIp(md.getPrivateIp());
+        instanceMetaDataEntry.setPublicIp(md.getPublicIp());
+        instanceMetaDataEntry.setSshPort(md.getSshPort());
+        instanceMetaDataEntry.setLocalityIndicator(md.getLocalityIndicator());
+    }
+
+    private void primaryGWSelectionFallbackIfNecessary(boolean primaryIgSelected, Map<String, InstanceGroup> instanceGroups) {
+        if (!primaryIgSelected) {
+            LOGGER.info("Primary GW was not selected! We will select one from the available GWs. It should not happen, it is a fallback mechanism");
+            selectPrimaryGWFromGatewayInstances(instanceGroups.values());
+        }
+    }
+
+    private boolean setupGatewayInstance(boolean primaryIgSelected, Optional<InstanceMetaData> terminatedPrimaryGwWhichShouldBeRestored,
+            InstanceMetaData instanceMetaDataEntry) {
+        LOGGER.info("Instance is a gateway instance: {}", instanceMetaDataEntry.getInstanceId());
+        if (terminatedPrimaryGwWhichShouldBeRestored.isPresent()) {
+            boolean primaryGWRestored = restorePrimaryGWIfFQDNMatch(terminatedPrimaryGwWhichShouldBeRestored.get(), instanceMetaDataEntry);
+            if (primaryGWRestored) {
+                primaryIgSelected = true;
```

Comment:
same here, please don't modify the parameter value, just insert return here
please add a log line